File: 2024/cpp/day6.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <set>
#include <vector>

using namespace std;

int part1();
int part2();

int main() {
  cout << part1() << endl;
  cout << part2() << endl;
}

int part1() {
  string line;
  ifstream input("../data/input6.txt");

  vector<string> maze;

  int start_x = 0, start_y = 0;

  if (input.is_open()) {

    int line_no = 0;
    while (getline(input, line)) {
      auto c = find(line.begin(), line.end(), '^');

      if (c != line.end()) {
        start_x = line_no;
        start_y = c - line.begin();
      }
      maze.push_back(line);

      line_no += 1;
    }
    input.close();
  }

  int height = maze.size();
  int width = maze[0].size();

  vector<pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
  int direction = 0;

  int x = start_x, y = start_y;

  set<pair<int, int>> visited;
  visited.insert({x, y});

  while (true) {
    int dx = directions[direction].first;
    int dy = directions[direction].second;

    int tx = x + dx;
    int ty = y + dy;

    if (!((tx < height && tx >= 0) && (ty < width && ty >= 0))) {
      break;
    }

    if (maze[tx][ty] == '#') {
      // rotate right 90deg
      direction = (direction + 1) % 4;
    } else {
      x = tx;
      y = ty;
      visited.insert({x, y});
    }
  }

  return visited.size();
}

int part2() {
  string line;
  ifstream input("../data/input6.txt");

  vector<string> maze;

  int start_x = 0, start_y = 0;

  if (input.is_open()) {

    int line_no = 0;
    while (getline(input, line)) {
      auto c = find(line.begin(), line.end(), '^');

      if (c != line.end()) {
        start_x = line_no;
        start_y = c - line.begin();
      }
      maze.push_back(line);

      line_no += 1;
    }
    input.close();
  }

  int height = maze.size();
  int width = maze[0].size();

  int loop_count = 0;

  for (int i = 0; i < height; i++) {
    for (int j = 0; j < width; j++) {

      if (maze[i][j] == '#')
        continue;
      if (maze[i][j] == '^')
        continue;

      maze[i][j] = '#';

      vector<pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
      int direction = 0;

      int x = start_x, y = start_y;

      set<pair<int, pair<int, int>>> visited;
      visited.insert({direction, {x, y}});

      while (true) {
        int dx = directions[direction].first;
        int dy = directions[direction].second;

        int tx = x + dx;
        int ty = y + dy;

        if (!((tx < height && tx >= 0) && (ty < width && ty >= 0))) {
          break;
        }

        if (maze[tx][ty] == '#') {
          // rotate right 90deg
          direction = (direction + 1) % 4;
        } else {
          x = tx;
          y = ty;

          if (visited.count({direction, {x, y}}) != 0) {
            loop_count++;
            break;
          }
          visited.insert({direction, {x, y}});
        }
      }

      maze[i][j] = '.';
    }
  }
  return loop_count;
}

File: 2024/python/day6.py
from collections import namedtuple


def part1() -> int:
    maze = []
    Point = namedtuple("Point", "x y")
    start = Point(0, 0)
    line_count = 0

    with open("../data/input6.txt") as inp:
        for line in inp.readlines():
            line_count += 1
            maze.append(list(line.strip()))
            if "^" in line:
                start = Point(line_count - 1, line.index("^"))

    visited_locs: set[Point] = set()
    visited_locs.add(start)

    maze_height = len(maze)
    maze_width = len(maze[0])

    obstacles: set[Point] = set(
        [
            Point(x, y)
            for x in range(maze_height)
            for y in range(maze_width)
            if maze[x][y] == "#"
        ]
    )

    current = start
    directions = (Point(-1, 0), Point(0, 1), Point(1, 0), Point(0, -1))
    direction = 0

    while True:
        if current.x >= len(maze) or current.x < 0:
            break

        if current.y >= len(maze[0]) or current.y < 0:
            break

        next_point = Point(
            current.x + directions[direction].x, current.y + directions[direction].y
        )

        if next_point in obstacles:
            # rotate 90deg right
            direction = (direction + 1) % 4
        else:
            # move forward
            current = next_point
            visited_locs.add(current)

    visited_locs.pop()
    return len(visited_locs)


def part2() -> int:
    maze = []
    Point = namedtuple("Point", "x y")
    start = Point(0, 0)
    line_count = 0

    with open("../data/input6.txt") as inp:
        for line in inp.readlines():
            line_count += 1
            maze.append(list(line.strip()))
            if "^" in line:
                start = Point(line_count - 1, line.index("^"))

    directions = (Point(-1, 0), Point(0, 1), Point(1, 0), Point(0, -1))
    maze_height = len(maze)
    maze_width = len(maze[0])
    loop_obs = 0

    for x, row in enumerate(maze):
        for y, c in enumerate(row):
            if c in "#^":
                continue

            maze[x][y] = "#"
            visited_locs = set()
            current = start
            direction = 0
            visited_locs.add((direction, start))

            while True:
                dx, dy = directions[direction]
                next_point = Point(current.x + dx, current.y + dy)

                if not (
                    0 <= next_point.x < maze_height and 0 <= next_point.y < maze_width
                ):
                    break

                if maze[next_point.x][next_point.y] == "#":
                    direction = (direction + 1) % 4
                else:
                    current = next_point
                    state = (direction, current)
                    if state in visited_locs:
                        loop_obs += 1
                        break
                    visited_locs.add(state)

            maze[x][y] = "."

    return loop_obs


print(part1())
# print(part2())
